File: src/main/java/virtualMachine/vm_instruction_processing/impl/PopInstructionProcessor.java
package virtualMachine.vm_instruction_processing.impl;

import virtualMachine.stack.memory.VirtualMemory;
import virtualMachine.stack.types.instruction.Instruction;
import virtualMachine.vm_instruction_processing.InstructionProcessor;

import static virtualMachine.stack.memory.MemorySegments.getFromName;

public class PopInstructionProcessor implements InstructionProcessor {

    @Override
    public void processInstruction(Instruction instruction, VirtualMemory virtualMemory) {
        var head = virtualMemory.pop();
        var reference = instruction.getNumericValue();
        var segment = getFromName(instruction.getOperand());

        virtualMemory.loadIntoMemory(head, reference, segment);
    }
}

File: src/main/kotlin/virtualMachine/stack/vm_instruction_parsing/vm_instruction_processing/impl/PopInstructionProcessor.kt
package virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.impl

import virtualMachine.stack.datawrappers.SixteenBit
import virtualMachine.stack.memory.MemorySegments
import virtualMachine.stack.memory.GlobalVirtualMemory
import virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.InstructionProcessor

class PopInstructionProcessor : InstructionProcessor {

    override fun processInstruction(instruction: String, virtualMemory: GlobalVirtualMemory) {

        val head : SixteenBit = virtualMemory.popLocalStack()
        val splitCommand: List<String> = instruction.split(" ")
        val reference = splitCommand[splitCommand.size - 1].toInt()
        val segment = MemorySegments.getFromName(splitCommand[1])

        virtualMemory.loadIntoMemory(head, reference, segment)
    }
}